### ISSUE TITLE
spack config update doesn't handle comments in yaml file

### ISSUE THREAD

**Issue by mwkrentel** [LLNL/spack#18031]
The `spack config update` command doesn't handle comments (#).

AFAIK, comments are allowed and I've been using them for years.  For
example, here is a sample `packages.yaml` file, and what happens when
I try to update it.

```
packages:
  # system cmake in /usr
  cmake:
    paths:
      cmake@3.11.0:  /usr
    buildable: False

$ spack config update packages
...
==> Do you want to proceed? [y/N] y
==> Error: 'dict' object has no attribute '_yaml_comment'
```

And the file isn't updated.

----------

Btw, there are two formats to specify lists in yaml, either with
`[...]` on one line, or with `-` on multiple lines.  I think this is
standard yaml and AFAIK, they're equivalent.

But spack update is inconsistent in how it updates them.  If I update
this, including cmake which needs updating:

```
packages:
  cmake:
    paths:
      cmake@3.11.0:  /usr
    buildable: False

  binutils:
    version:  [2.34]
    variants: +libiberty ~nls
```

Then binutils changes to:

```
  binutils:
    version:
    - 2.34
    variants: +libiberty ~nls
```

But if I leave out cmake, then spack update reports nothing to do and
leaves the file unchanged.

```
packages:
  binutils:
    version:  [2.34]
    variants: +libiberty ~nls

$ spack config update packages
==> No updates needed for "packages" section.
```

I don't know if I really care one way or the other, but it's odd that
it's inconsistent like this.

----------

Also, spack update deletes all the blank lines in the file, which I
don't like.

**Comment by alalazo** [comment]
Bug confirmed:
```console
$ cat ~/.spack/packages.yaml
packages:
  # system cmake in /usr
  cmake:
    paths:
      cmake@3.11.0:  /usr
    buildable: False

$ spack -d config update packages -y
==> [2020-08-13-08:29:36.695023] Imported config from built-in commands
==> [2020-08-13-08:29:36.700339] Imported config from built-in commands
==> [2020-08-13-08:29:36.701393] Reading config file /home/culpo/PycharmProjects/spack/etc/spack/defaults/packages.yaml
==> [2020-08-13-08:29:36.724921] Reading config file /home/culpo/.spack/packages.yaml
==> [2020-08-13-08:29:36.728661] Warning: the attribute "paths" in the "packages" section of the configuration has been deprecated [entry=CommentedMap([('cmake@3.11.0', '/usr')])]
==> [2020-08-13-08:29:36.728953] OUTDATED CONFIGURATION FILE [section=packages, scope=user, dir=/home/culpo/.spack]
==> [2020-08-13-08:29:36.730702] Reading config file /home/culpo/PycharmProjects/spack/etc/spack/defaults/config.yaml
Traceback (most recent call last):
  File "/home/culpo/PycharmProjects/spack/bin/spack", line 64, in <module>
    sys.exit(spack.main.main())
  File "/home/culpo/PycharmProjects/spack/lib/spack/spack/main.py", line 762, in main
    return _invoke_command(command, parser, args, unknown)
  File "/home/culpo/PycharmProjects/spack/lib/spack/spack/main.py", line 490, in _invoke_command
    return_val = command(parser, args)
  File "/home/culpo/PycharmProjects/spack/lib/spack/spack/cmd/config.py", line 455, in config
    action[args.config_command](args)
  File "/home/culpo/PycharmProjects/spack/lib/spack/spack/cmd/config.py", line 374, in config_update
    args.section, data, scope=scope.name, force=True
  File "/home/culpo/PycharmProjects/spack/lib/spack/spack/config.py", line 490, in update_config
    comments)
AttributeError: 'dict' object has no attribute '_yaml_comment'
```
Currently using :

* **Spack:** 0.15.3-508-5512340a5
* **Python:** 3.6.9
* **Platform:** linux-ubuntu18.04-broadwell

**Comment by alalazo** [comment]
@mwkrentel Thanks for reporting this bug. There's a fix for it in #18045 Let me know if it works for you. 

---

Concerning the style issues, to know if we need to update a file we check its content in memory. If we have to write updated content in it we rely on the `ruamel.yaml`  we vendor for formatting and I think currently we don't tweak much defaults. I'd avoid to go too deep into formatting, since it might be an opinionated topic and it might require quite some effort to track different sections of a file (what if somebody uses flow style for a sequence and block style for another sequence in the same YAML?)

**Comment by mwkrentel** [comment]
@alalazo  Well, it doesn't crash, which is good.

The following example is still not handled well, where the top and
bottom block of comments are not copied to the new version.

```
#
# packages.yaml file
# these comments get erased
#
packages:

  # system cmake in /usr
  cmake:
    paths:
      cmake@3.11.0:  /usr
    buildable: False

  # example how to build binutils
  # these comments are also erased
  #
  # binutils:
  #   version:  [2.34]
  #   variants: +libiberty ~nls
```

I can tell by how spack reads things in and then prints the results
that the file gets "normalized" in the internal representation.  So,
it's probably impossible to transform the file the way a person would,
and not worth trying to do so.

So, I'm willing to say, "close enough," and move on.
